core/assignment.go

```diff
@@ -149,38 +143,101 @@ func (c *StdAssignmentCoordinator) GetOperatorAssignment(state *OperatorState, q
+	info := header.QuorumInfos[quorum]
+
+	// Check that the chunk length meets the minimum requirement
+	if info.ChunkLength < minChunkLength {
+		return false, ErrChunkLengthTooSmall
 	}
 
-	numSys := roundUpDivide(uint(quorumThreshold-adversaryThreshold)*numOperators*quantizationFactor, PercentMultiplier)
-	chunkLength := roundUpDivide(blobLength, numSys)
-	return chunkLength, nil
+	// Get minimum stake amont
+	minStake := state.Totals[quorum].Stake
+	for _, r := range state.Operators[quorum] {
+		if r.Stake.Cmp(minStake) < 0 {
+			minStake = r.Stake
+		}
+	}
+
+	totalStake := state.Totals[quorum].Stake
+	if info.ChunkLength != minChunkLength {
+
+		num := new(big.Int).Mul(big.NewInt(2*int64(header.Length*percentMultiplier)), minStake)
+		denom := new(big.Int).Mul(big.NewInt(int64(info.QuorumThreshold-info.AdversaryThreshold)), totalStake)
+		maxChunkLength := uint(roundUpDivideBig(num, denom).Uint64())
+
+		maxChunkLength2 := roundUpDivide(2*header.Length*percentMultiplier, maxRequiredNumChunks*uint(info.QuorumThreshold-info.AdversaryThreshold))
+
+		if maxChunkLength < maxChunkLength2 {
+			maxChunkLength = maxChunkLength2
+		}
+
+		maxChunkLength = uint(nextPowerOf2(uint64(maxChunkLength)))
+
+		if info.ChunkLength > maxChunkLength {
+			return false, ErrChunkLengthTooLarge
+		}
+
+	}
+
+	return true, nil
 
 }
 
-func (c *StdAssignmentCoordinator) GetChunkLengthFromHeader(state *OperatorState, header *BlobQuorumInfo) (uint, error) {
+func (c *StdAssignmentCoordinator) CalculateChunkLength(state *OperatorState, blobLength, targetNumChunks uint, param *SecurityParam) (uint, error) {
+
+	quorum := param.QuorumID
+	numOperators := len(state.Operators[quorum])
+
+	var numChunks uint
+	if targetNumChunks != 0 {
+		// If targetNumChunks is specified, then we use that value
+
+		numChunks = targetNumChunks - uint(numOperators)
+
+	} else {
+		// Otherwise, we calculate the number of chunks based on the ratio of the total stake to the minimum stake
+
+		// Get minimum stake amount
+		minStake := state.Totals[quorum].Stake
+		for _, r := range state.Operators[quorum] {
+			if r.Stake.Cmp(minStake) < 0 {
+				minStake = r.Stake
+			}
+		}
+
+		// Cap the number of chunks to maxRequiredNumChunks - numOperators
+		numChunks = maxRequiredNumChunks - uint(numOperators)
 
-	// Validate the chunk length
-	numOperators := uint(len(state.Operators[header.QuorumID]))
-	chunkLength := header.EncodedBlobLength / (header.QuantizationFactor * numOperators)
+		if minStake.Cmp(big.NewInt(0)) > 0 {
+			totalStake := state.Totals[quorum].Stake
+			numChunksByStake := uint(roundUpDivideBig(totalStake, minStake).Uint64())
+
+			if numChunksByStake < numChunks {
+				numChunks = numChunksByStake
+			}
+		}
 
-	if chunkLength*header.QuantizationFactor*numOperators != header.EncodedBlobLength {
-		return 0, errors.New("invalid header")
 	}
 
+	chunkLength := roundUpDivide(blobLength*percentMultiplier, numChunks*uint(param.QuorumThreshold-param.AdversaryThreshold))
```

Comment:
So chunkLength is depending on numChunks here, but in `GetAssignments' (L97), numChunks is depending on chunkLength. How is there a cyclic dependency in determining these two parameter? Or is it sort of iterative?

Reply:
Please take a look at the updated function with the comment. I think you might find this easier to understand.